### PR TITLE
test: fix remaining msan errors.

### PR DESCRIPTION
**Pull request by htuch**
These manifested as msan errors (use of uninitialized memory) in
ssl_integration_test, but were actually the result of a combination of
racey behavior and timeouts due to slower test execution with
instrumentation and loaded test machines.

This patch makes the following changes:

* Increase all timeouts in integration test JSON to 5 seconds, to avoid
  flaking on slow remote test servers or under additional instrumentation.

* Fix race condition in FakeUpstream, where waitFor{Http,Raw}Connection
  would consume a Network::Connection that might have been freed already
  (due to a client disconnection).

* Harden FakeConnectionBase::readDisable() to work across disconnects.

### CODE REVIEW

File: test/integration/fake_upstream.cc [hunk 1]
```diff
@@ -147,7 +147,18 @@ void FakeConnectionBase::close() {
 }
 
 void FakeConnectionBase::readDisable(bool disable) {
-  connection_.dispatcher().post([this, disable]() -> void { connection_.readDisable(disable); });
+  std::unique_lock<std::mutex> lock(lock_);
+  if (!disconnected_) {
+    connection_.dispatcher().post([this, disable]() -> void {
+      if (!disconnected_) {
+        connection_.readDisable(disable);
+      }
```
Comment (mattklein123) [inline test/integration/fake_upstream.cc]: Should we put an else clause here with NOT_IMPLEMENTED? Seems like race could still happen? Or can that one happen for real?

File: test/integration/integration.cc [hunk 2]
```diff
@@ -813,6 +813,7 @@ void BaseIntegrationTest::testUpstreamProtocolError() {
   IntegrationCodecClientPtr codec_client;
   IntegrationStreamDecoderPtr response(new IntegrationStreamDecoder(*dispatcher_));
   FakeRawConnectionPtr fake_upstream_connection;
+  fake_upstreams_[0]->setWaitForHttpConnection(false);
```
Comment (mattklein123) [inline test/integration/integration.cc]: I'm a little confused by this part of the change. It seems fragile that we have to remember to set this outside the test infra. Why is this needed? And we still call makeHttpConnection() below?

File: test/integration/fake_upstream.cc (hunk 1 above)
Comment (htuch) [inline test/integration/fake_upstream.cc]: Yeah, sure will add.

File: test/integration/integration.cc (hunk 2 above)
Comment (htuch) [inline test/integration/integration.cc]: We need `FakeUpstream` to be capable of asynchronously waiting for and queuing up arrived connections in `new_connections_` (or the variants I added in this PR).  Previously, the `Network::Connection&` was queued up here, but it was not owned and could be freed on a disconnect. The consumer of `new_connection_` then gets some freed memory. There's no event listener to catch this scenario, that's why I want to wrap the connection with one of these `Fake*Connection` objects that is capable of responding to the disconnect if it happens while queued in `new_connections_`.

Maybe we could get `FakeHttpConnection` to wrap around a `FakeRawConnection`, I can look into this. It seemed to me the common case that folks want to `waitForHttpConnection()` though, only a few care about `waitForRawConnection()`.

Comment (mattklein123) [inline test/integration/integration.cc]: Talked offline and I think we have a slightly better path forward on this.

**Comment by htuch** [comment]
Updated with connection wrapper, did a full squash/push -f to forget about the previous solution.

File: test/integration/fake_upstream.cc (hunk 1 above)
Comment (mattklein123) [inline test/integration/fake_upstream.cc]: I think the changes in this function are still trying to fix a previous race condition. Should we remove them fro this PR or just RELEASE_ASSERT(!disconnected_) after taking the lock?

File: test/integration/fake_upstream.h [hunk 3]
```diff
@@ -70,6 +70,35 @@ class FakeStream : public Http::StreamDecoder, public Http::StreamCallbacks {
 typedef std::unique_ptr<FakeStream> FakeStreamPtr;
 
 /**
+ * Wraps a raw Network::Connection in a safe way, such that the connection can
+ * be placed in a queue for an arbitrary amount of time. It handles disconnects
+ * that take place in the queued state by failing the test. Once a
+ * QueuedConnectionWrapper object is instantiated by FakeHttpConnection or
+ * FakeRawConnection, it no longer plays a role.
```
Comment (mattklein123) [inline test/integration/fake_upstream.h]: nit: I would add a TODO here that if we add remove callback support from connection we can simplify this.
